plugins/poll/assets/javascripts/widgets/discourse-poll.js.es6

```diff
@@ -415,6 +422,210 @@ createWidget("discourse-poll-info", {
   }
 });
 
+function fetchGroupedVoters(data) {
+  return ajax("/polls/grouped_poll_results.json", { data }).catch(error => {
+    if (error) {
+      popupAjaxError(error);
+    } else {
+      bootbox.alert(I18n.t("poll.error_while_fetching_voters"));
+    }
+  });
+}
+
+function transformUserFieldToLabel(fieldName) {
+  let newName = fieldName;
+  return newName.replace(/_/g, " ");
+}
+
+createWidget("discourse-poll-grouped-pies", {
+  tagName: "div.poll-grouped-pies",
+  buildAttributes(attrs) {
+    return {
+      id: `poll-results-grouped-pie-charts-${attrs.id}`
+    };
+  },
+
+  html(attrs) {
+    const fields = Object.assign({}, attrs.groupableUserFields);
+    later(() => {
+      attrs.groupedBy = attrs.groupedBy || fields[0];
+      const parent = document.getElementById(
+        `poll-results-grouped-pie-charts-${attrs.id}`
+      );
+      const fieldSelectId = `field-select-${attrs.id}`;
+
+      const existingSel = document.getElementById(fieldSelectId);
+      if (existingSel) {
+        existingSel.parentNode.removeChild(existingSel);
+      }
+
+      const sel = document.createElement("select");
+      sel.classList.add("poll-group-by-selector");
+      sel.id = fieldSelectId;
+
+      const clearFix = document.createElement("div");
+      clearFix.classList.add("clear-both");
+
+      parent.prepend(sel);
+      parent.append(clearFix);
+
+      for (var i = 0; i < attrs.groupableUserFields.length; i++) {
+        const opt = document.createElement("option");
+        opt.innerHTML = transformUserFieldToLabel(fields[i]);
+        opt.value = fields[i];
+        sel.appendChild(opt);
+        $(sel).val(attrs.groupedBy);
+      }
+      fetchGroupedVoters({
+        post_id: attrs.post.id,
+        poll_name: attrs.poll.name,
+        user_field_name: attrs.groupedBy
+      }).then(result => {
+        for (
```

Comment:
I see you use `for` a lot, I didn't look at each of them, but might be worth considering using `forEach`